google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/admin/v2/models/GCRules.java

```diff
@@ -0,0 +1,315 @@
+        default:
+          return GcRule.newBuilder().setUnion(builder.build()).build();
+      }
+    }
+  }
+
+  /** Wrapper for building max versions rule */
+  public static final class VersionRule extends BaseRule {
+    private GcRule.Builder builder;
+
+    private VersionRule(int maxVersion) {
+      this.builder = GcRule.newBuilder();
+      builder.setMaxNumVersions(maxVersion);
+    }
+
+    /** Gets the configured maximum versions */
+    public int getMaxVersions() {
+      return toProto().getMaxNumVersions();
+    }
+
+    @Override
+    public String toString() {
+      return MoreObjects.toStringHelper(this).add("maxNumVersions", getMaxVersions()).toString();
+    }
+
+    @InternalApi
+    @Override
+    public GcRule toProto() {
+      return builder.build();
+    }
+  }
+
+  /** Wrapper for building max duration rule */
+  public static final class DurationRule extends BaseRule {
+    private com.google.protobuf.Duration.Builder builder;
+
+    private DurationRule(Duration duration) {
+      this.builder =
+          com.google.protobuf.Duration.newBuilder()
+              .setSeconds(duration.getSeconds())
+              .setNanos(duration.getNano());
+    }
+
+    /**
+     * Gets the configured maximum age
+     *
+     * @return Duration
+     */
+    public Duration getMaxAge() {
+      long seconds = toProto().getMaxAge().getSeconds();
+      int nanos = toProto().getMaxAge().getNanos();
+      return Duration.ofSeconds(seconds, nanos);
+    }
+
+    @Override
+    public String toString() {
+      return MoreObjects.toStringHelper(this).add("maxAge", getMaxAge()).toString();
+    }
+
+    @InternalApi
+    @Override
+    public GcRule toProto() {
+      return GcRule.newBuilder().setMaxAge(builder.build()).build();
+    }
+  }
+
+  /** Wrapper for building a empty rule */
+  public static final class DefaultRule extends BaseRule {
+    private DefaultRule() {}
+
+    @InternalApi
+    @Override
+    public GcRule toProto() {
+      return GcRule.getDefaultInstance();
+    }
+
+    @Override
+    public String toString() {
+      return MoreObjects.toStringHelper(this).toString();
+    }
+  }
+
+  /** Helpers to support casting rules the Type, when the Type is known */
+  public abstract static class BaseRule implements GCRule {
+
+    /** Casts the rule to DurationRule. On failure throws a ClassCastException */
+    public DurationRule getDurationOrThow() {
+      return (DurationRule) this;
+    }
+
+    /** Casts the rule to VersionRule. On failure throws a ClassCastException */
+    public VersionRule getVersionOrThow() {
+      return (VersionRule) this;
+    }
+
+    /** Casts the rule to UnionRule. On failure throws a ClassCastException */
+    public UnionRule getUnionOrThow() {
+      return (UnionRule) this;
+    }
+
+    /** Casts the rule to IntersectionRule. On failure throws a ClassCastException */
+    public IntersectionRule getIntersectionOrThow() {
+      return (IntersectionRule) this;
+    }
```

Comment:
I don't think these OrThrow methods are necessary. The user can cast it themselves

Reply:
I'm not sure what you mean:

```java
((IntersectionRule)rule)

// vs
rule.getIntersectionOrThow()
```

I really don't like having the base class knowing about its subclasses.